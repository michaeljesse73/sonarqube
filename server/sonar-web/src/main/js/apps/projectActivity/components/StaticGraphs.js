/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import moment from 'moment';
import { some, sortBy } from 'lodash';
import { AutoSizer } from 'react-virtualized';
import AdvancedTimeline from '../../../components/charts/AdvancedTimeline';
import StaticGraphsLegend from './StaticGraphsLegend';
import { formatMeasure, getShortType } from '../../../helpers/measures';
import { EVENT_TYPES } from '../utils';
import { translate } from '../../../helpers/l10n';
import type { Analysis } from '../types';
import type { Serie } from '../../../components/charts/AdvancedTimeline';

type Props = {
  analyses: Array<Analysis>,
  eventFilter: string,
  graphEndDate: ?Date,
  graphStartDate: ?Date,
  leakPeriodDate: Date,
  loading: boolean,
  metricsType: string,
  series: Array<Serie>,
  showAreas?: boolean,
  updateGraphZoom: (from: ?Date, to: ?Date) => void
};

export default class StaticGraphs extends React.PureComponent {
  props: Props;

  formatYTick = tick => formatMeasure(tick, getShortType(this.props.metricsType));

  getEvents = () => {
    const { analyses, eventFilter } = this.props;
    const filteredEvents = analyses.reduce((acc, analysis) => {
      if (analysis.events.length <= 0) {
        return acc;
      }
      let event;
      if (eventFilter) {
        event = analysis.events.filter(event => event.category === eventFilter)[0];
      } else {
        event = sortBy(analysis.events, event => EVENT_TYPES.indexOf(event.category))[0];
      }
      if (!event) {
        return acc;
      }
      return acc.concat({
        className: event.category,
        name: event.name,
        date: moment(analysis.date).toDate()
      });
    }, []);
    return sortBy(filteredEvents, 'date');
  };

  hasSeriesData = () => some(this.props.series, serie => serie.data && serie.data.length > 2);

  render() {
    const { loading } = this.props;

    if (loading) {
      return (
        <div className="project-activity-graph-container">
          <div className="text-center">
            <i className="spinner" />
          </div>
        </div>
      );
    }

    if (!this.hasSeriesData()) {
      return (
        <div className="project-activity-graph-container">
          <div className="note text-center">
            {translate('component_measures.no_history')}
          </div>
        </div>
      );
    }

    const { series } = this.props;
    return (
      <div className="project-activity-graph-container">
        <StaticGraphsLegend series={series} />
        <div className="project-activity-graph">
          <AutoSizer>
            {({ height, width }) => (
              <AdvancedTimeline
                endDate={this.props.graphEndDate}
                events={this.getEvents()}
                height={height}
                width={width}
                interpolate="linear"
                formatYTick={this.formatYTick}
                leakPeriodDate={this.props.leakPeriodDate}
                metricType={this.props.metricsType}
                series={series}
                showAreas={this.props.showAreas}
                startDate={this.props.graphStartDate}
                updateZoom={this.props.updateGraphZoom}
              />
            )}
          </AutoSizer>
        </div>
      </div>
    );
  }
}
