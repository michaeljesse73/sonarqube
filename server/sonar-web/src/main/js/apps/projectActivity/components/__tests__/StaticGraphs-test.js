/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import { shallow } from 'enzyme';
import StaticGraphs from '../StaticGraphs';

const ANALYSES = [
  {
    key: 'A1',
    date: new Date('2016-10-27T16:33:50+0200'),
    events: [
      {
        key: 'E1',
        category: 'VERSION',
        name: '6.5-SNAPSHOT'
      }
    ]
  },
  {
    key: 'A2',
    date: new Date('2016-10-27T12:21:15+0200'),
    events: []
  },
  {
    key: 'A3',
    date: new Date('2016-10-26T12:17:29+0200'),
    events: [
      {
        key: 'E2',
        category: 'OTHER',
        name: 'foo'
      },
      {
        key: 'E3',
        category: 'VERSION',
        name: '6.4'
      }
    ]
  }
];

const SERIES = [
  {
    name: 'bugs',
    translatedName: 'metric.bugs.name',
    style: 0,
    data: [
      { x: new Date('2016-10-27T16:33:50+0200'), y: 5 },
      { x: new Date('2016-10-27T12:21:15+0200'), y: 16 },
      { x: new Date('2016-10-26T12:17:29+0200'), y: 12 }
    ]
  }
];

const EMPTY_SERIES = [
  {
    name: 'bugs',
    translatedName: 'metric.bugs.name',
    style: 0,
    data: []
  }
];

const DEFAULT_PROPS = {
  analyses: ANALYSES,
  eventFilter: '',
  filteredSeries: SERIES,
  leakPeriodDate: '2017-05-16T13:50:02+0200',
  loading: false,
  series: SERIES,
  metricsType: 'INT'
};

it('should show a loading view', () => {
  expect(shallow(<StaticGraphs {...DEFAULT_PROPS} loading={true} />)).toMatchSnapshot();
});

it('should show that there is no data', () => {
  expect(shallow(<StaticGraphs {...DEFAULT_PROPS} series={EMPTY_SERIES} />)).toMatchSnapshot();
});

it('should correctly render a graph', () => {
  expect(shallow(<StaticGraphs {...DEFAULT_PROPS} />)).toMatchSnapshot();
});

it('should correctly filter events', () => {
  expect(shallow(<StaticGraphs {...DEFAULT_PROPS} />).instance().getEvents()).toMatchSnapshot();
  expect(
    shallow(<StaticGraphs {...DEFAULT_PROPS} eventFilter="OTHER" />).instance().getEvents()
  ).toMatchSnapshot();
});
