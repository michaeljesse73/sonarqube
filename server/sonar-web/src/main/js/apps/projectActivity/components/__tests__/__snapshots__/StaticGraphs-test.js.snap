// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly filter events 1`] = `
Array [
  Object {
    "className": "VERSION",
    "date": 2016-10-26T10:17:29.000Z,
    "name": "6.4",
  },
  Object {
    "className": "VERSION",
    "date": 2016-10-27T14:33:50.000Z,
    "name": "6.5-SNAPSHOT",
  },
]
`;

exports[`should correctly filter events 2`] = `
Array [
  Object {
    "className": "OTHER",
    "date": 2016-10-26T10:17:29.000Z,
    "name": "foo",
  },
]
`;

exports[`should correctly render a graph 1`] = `
<div
  className="project-activity-graph-container"
>
  <StaticGraphsLegend
    series={
      Array [
        Object {
          "data": Array [
            Object {
              "x": 2016-10-27T14:33:50.000Z,
              "y": 5,
            },
            Object {
              "x": 2016-10-27T10:21:15.000Z,
              "y": 16,
            },
            Object {
              "x": 2016-10-26T10:17:29.000Z,
              "y": 12,
            },
          ],
          "name": "bugs",
          "style": 0,
          "translatedName": "metric.bugs.name",
        },
      ]
    }
  />
  <div
    className="project-activity-graph"
  >
    <AutoSizer
      onResize={[Function]}
    />
  </div>
</div>
`;

exports[`should show a loading view 1`] = `
<div
  className="project-activity-graph-container"
>
  <div
    className="text-center"
  >
    <i
      className="spinner"
    />
  </div>
</div>
`;

exports[`should show that there is no data 1`] = `
<div
  className="project-activity-graph-container"
>
  <div
    className="note text-center"
  >
    component_measures.no_history
  </div>
</div>
`;
